grr/server/grr_response_server/gui/ui/lib/api_translation/primitive_test.ts

```diff
@@ -70,6 +70,65 @@ describe('createOptionalDateSeconds', () => {
   });
   it('handles unixtimes correctly', () => {
     expect(createOptionalDateSeconds('1579167695'))
-        .toEqual(new Date(1579167695000));
+      .toEqual(new Date(1579167695000));
+  });
+});
+
+describe('decodeBase64', () => {
+  it('returns empty byte array for undefined', () => {
+    expect(decodeBase64(undefined)).toEqual(new Uint8Array());
+  });
+  it('returns empty byte array for invalid base64 encoded string ', () => {
+    expect(decodeBase64('Inv@lid $tring')).toEqual(new Uint8Array());
+  });
+  it('returns correct byte array for the base64 encoded string ', () => {
+    expect(decodeBase64('yv66vg==')).toEqual(new Uint8Array([0xCA, 0xFE, 0xBA, 0xBE]));
+  });
+});
+
+describe('byteToHex', () => {
+  it('translates only the first byte', () => {
+    expect(byteToHex(0xBABE)).toEqual('BE');
+  });
+  it('adds 0 padding on front', () => {
+    expect(byteToHex(0x1)).toEqual('01');
+  });
+  it('adds 0 padding on front', () => {
+    expect(byteToHex(0xA)).toEqual('0A');
+  });
+});
+
+describe('createIpv4Address', () => {
+  it('creates an empty string on non 4 bytes array input', () => {
+    expect(createIpv4Address(new Uint8Array())).toEqual('');
+    expect(createIpv4Address(new Uint8Array([1, 2, 3]))).toEqual('');
+    expect(createIpv4Address(new Uint8Array([1, 2, 3, 4, 5]))).toEqual('');
+  });
+  it('creates the IPv4 address representation of a 4 bytes array', () => {
+    expect(createIpv4Address(new Uint8Array([1, 2, 3, 4]))).toEqual('1.2.3.4');
+    expect(createIpv4Address(new Uint8Array([127, 0, 0, 1]))).toEqual('127.0.0.1');
+  });
+});
+
+describe('createIpv6Address', () => {
+  it('creates an empty string on non 16 bytes array input', () => {
+    expect(createIpv6Address(new Uint8Array())).toEqual('');
+    expect(createIpv6Address(new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]))).toEqual('');
+    expect(createIpv6Address(new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17]))).toEqual('');
+  });
+  it('creates the IPv6 non-abbreviated address representation of a 16 bytes array', () => {
+    expect(createIpv6Address(new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]))).toEqual('0102:0304:0506:0708:090A:0B0C:0D0E:0F10');
+  });
+});
+
+describe('createMacAddress', () => {
+  it('creates an empty string on non 6 bytes array input', () => {
+    expect(createMacAddress(new Uint8Array())).toEqual('');
+    expect(createMacAddress(new Uint8Array([1, 2, 3, 4, 5,]))).toEqual('');
+    expect(createMacAddress(new Uint8Array([1, 2, 3, 4, 5, 6, 7]))).toEqual('');
+  });
+  it('creates the MAC address representation of a 6 bytes array', () => {
+    expect(createMacAddress(new Uint8Array([1, 2, 3, 4, 5, 6]))).toEqual('01:02:03:04:05:06');
+    expect(createMacAddress(new Uint8Array([10, 11, 12, 13, 14, 15]))).toEqual('0A:0B:0C:0D:0E:0F');
```

Comment:
Please add more test cases, especially how the mad and ip code handle intermediate blocks of zeroes and minimum and maximum values.